rasa/utils/tensorflow/model_data_utils.py

```diff
@@ -86,79 +92,77 @@ def create_zero_features(features: List[List[List["Features"]]],) -> List["Featu
-        features = [[dicts] for dicts in features]
+    if isinstance(tracker_state_features[0], Dict):
+        tracker_state_features = [[dicts] for dicts in tracker_state_features]
         remove_sequence_dimension = True
 
-    features = surface_attributes(features)
+    state_to_tracker_features = surface_attributes(tracker_state_features)
 
     attribute_data = {}
 
     # During prediction we need to iterate over the zero features attributes to
     # have all keys in the resulting model data
     if training:
-        attributes = list(features.keys())
+        attributes = list(state_to_tracker_features.keys())
     else:
-        attributes = list(zero_features.keys())
+        attributes = list(zero_state_features.keys())
 
+    # In case an attribute is not present during prediction, replace it with
+    # None values that will then be replaced by zero features
     dialogue_length = 1
-    for key, values in features.items():
-        dialogue_length = max(dialogue_length, len(values[0]))
-
+    for tracker_features in state_to_tracker_features.values():
+        dialogue_length = max(dialogue_length, len(tracker_features[0]))
     empty_features = [[None] * dialogue_length]
 
     for attribute in attributes:
-        features_in_tracker = (
-            features[attribute] if attribute in features else empty_features
+        tracker_features = (
+            state_to_tracker_features[attribute]
+            if attribute in state_to_tracker_features
+            else empty_features
         )
 
         # in case some features for a specific attribute and dialogue turn are
         # missing, replace them with a feature vector of zeros
         if training:
-            zero_features[attribute] = create_zero_features(features_in_tracker)
+            zero_state_features[attribute] = create_zero_features(tracker_features)
 
         (attribute_masks, _dense_features, _sparse_features,) = map_tracker_features(
-            features_in_tracker, zero_features[attribute]
+            tracker_features, zero_state_features[attribute]
         )
 
         sparse_features = defaultdict(list)
         dense_features = defaultdict(list)
 
-        if remove_sequence_dimension:
-            # remove added sequence dimension
-            for key, values in _sparse_features.items():
-                sparse_features[key] = [value[0] for value in values]
-            for key, values in _dense_features.items():
-                dense_features[key] = [value[0] for value in values]
-        else:
-            for key, values in _sparse_features.items():
-                sparse_features[key] = [scipy.sparse.vstack(value) for value in values]
-            for key, values in _dense_features.items():
-                dense_features[key] = [np.vstack(value) for value in values]
-
-        # TODO not sure about expand_dims
+        # vstack serves as removing dimension
+        # TODO check vstack for sequence features

```

Comment:
for sequence features `vstack ` won't work just like that, but we'll have to change a little bit how we pad a batches in RasaModelData to manage 4d tensors.

Reply:
sure, I left a `TODO` so that we don't forget